src/scarabee/_scarabee/cmfd.cpp

```diff
@@ -325,26 +477,788 @@ void CMFD::compute_homogenized_xs_and_flux(const MOCDriver& moc) {
+    // Loop over MOC groups
+    for (std::size_t g = 0; g < moc_to_cmfd_group_map_.size(); g++) {
+      // Loop over FSRs in cell l
+      for (auto fsr : fsrs_[l]) {
+        extern_src_[moc_to_cmfd_group_map_[g] * tot_cells + l] +=
+            invs_V * moc.volume(fsr) * moc.extern_src(fsr, g);
+      }
+    }
+  }
+}
+
+void CMFD::check_neutron_balance(const std::size_t i, const std::size_t j,
+                                 std::size_t g, const double keff) const {
+  // First, we get the spacing of the tile
+  const double dx = dx_[i];
+  const double dy = dy_[j];
+
+  // Next, get the surfaces for out tile
+  const double J_xn = surface_currents_.at(g, j * x_bounds_.size() + i);
+  const double J_xp = surface_currents_.at(g, j * x_bounds_.size() + i + 1);
+  const double J_yn =
+      surface_currents_.at(g, nx_surfs_ + i * y_bounds_.size() + j);
+  const double J_yp =
+      surface_currents_.at(g, nx_surfs_ + i * y_bounds_.size() + j + 1);
+
+  // Get the xs for our tile
+  const auto& xs = *xs_(i, j);
+  const double chi_g_keff = xs.chi(g) / keff;
+
+  // Compute the fission and scatter sources
+  double fiss_source = 0.;
+  double scat_source = 0.;
+  for (std::size_t gg = 0; gg < ng_; gg++) {
+    const double flx_gg = flux_.at(gg, i, j);
+    fiss_source += chi_g_keff * xs.vEf(gg) * flx_gg;
+    scat_source += xs.Es(gg, g) * flx_gg;
+  }
+
+  // Compute the total leakage from the currents
+  const double leak_rate = ((J_xp - J_xn) / dx) + ((J_yp - J_yn) / dy);
+
+  // Now compute the removal reaction rate
+  const double tot_reac_rate = Et_(g, i, j) * flux_.at(g, i, j);
+
+  // Compute the residual of the balance equation
+  const double residual =
+      leak_rate + tot_reac_rate - (scat_source + fiss_source);
+
+  if (std::abs(residual) >= 1.E-5) {
+    spdlog::error(
+        "CMFD tile ({:d}, {:d}) in group {:d} has a neutron balance residual "
+        "of "
+        "{:.5E}.",
+        i, j, g, residual);
+  }
+}
+
+void CMFD::set_damping(double wd) {
+  if (wd < 0.0 || wd > 1.0) {
+    auto mssg = "Damping factor for CMFD must be in interval [0.0, 1.0].";
+    spdlog::error(mssg);
+    throw ScarabeeException(mssg);
+  }
+  damping_ = wd;
+}
+
+void CMFD::set_flux_tolerance(double ftol) {
+  if (ftol <= 0. || ftol >= 0.1) {
+    auto mssg = "Tolerance for flux must be in the interval (0., 0.1).";
+    spdlog::error(mssg);
+    throw ScarabeeException(mssg);
+  }
+
+  flux_tol_ = ftol;
+}
+
+void CMFD::set_keff_tolerance(double ktol) {
+  if (ktol <= 0. || ktol >= 0.1) {
+    auto mssg = "Tolerance for keff must be in the interval (0., 0.1).";
+    spdlog::error(mssg);
+    throw ScarabeeException(mssg);
+  }
+
+  keff_tol_ = ktol;
+}
+
+void CMFD::set_larsen_correction(bool user_pref) {
+  if (od_cmfd_) {
+    auto mssg =
+        "odCMFD and Larsen Correction options are mutally exclusive. Disable "
+        "odCMFD with 'od_cmfd = False'.";
+    spdlog::error(mssg);
+    throw ScarabeeException(mssg);
+  }
```

Comment:
This logic isn't quite correct. The user_pref could be false !
```suggestion
  if (od_cmfd_ && user_pref) {
    auto mssg =
        "odCMFD and Larsen Correction options are mutally exclusive. Disable "
        "odCMFD with 'od_cmfd = False'.";
    spdlog::error(mssg);
    throw ScarabeeException(mssg);
  }
```